docs/architecture/adr-001-autopilot-oci-basic-auth-support.md

```diff
@@ -0,0 +1,127 @@
+# ADR 1: Add Support for OCI Registry and HTTP Authentication
+
+## Context
+
+Registries are increasingly being used as generic artifact stores, expanding beyond their traditional role of hosting container images. To align with this trend, it is beneficial for Autopilot to support pulling artifacts directly from registries. Currently, Autopilot's capabilities are limited to downloading artifacts via the HTTP[S] protocols.
+
+Enhancing Autopilot to pull artifacts directly from registries will streamline workflows and improve efficiency by allowing integration and deployment of diverse artifacts without relying solely on HTTP[S] endpoints. This update will enable Autopilot to handle registry-specific protocols and authentication mechanisms, aligning it with modern deployment practices.
+
+Currently, Autopilot does not support the retrieval of artifacts via the HTTP protocol when authentication is required. Implementing this feature to accommodate such authentication methods would be a valuable enhancement.
+
+## Decision
+
+Implement support in Autopilot for pulling artifacts, such as k0s binaries and image bundles, directly from a registry using the [ORAS](https://oras.land/docs/) client. Additionally, add support for HTTP authentication to ensure secure access to artifacts.
+
+## Solution
+
+Starting with the current `PlanResourceURL` struct:
+
+```go
+type PlanResourceURL struct {
+        // URL is the URL of a downloadable resource.
+        URL string `json:"url"`
+
+        // Sha256 provides an optional SHA256 hash of the URL's content for verification.
+        Sha256 string `json:"sha256,omitempty"`
+}
+```
+
+We must specify to Autopilot where to access credentials for remote artifact pulls. This will be achieved by adjusting the struct as follows:
+
+```go
+type PlanResourceURL struct {
+        // URL is the URL of a downloadable resource.
+        URL string `json:"url"`
+
+        // Sha256 provides an optional SHA256 hash of the URL's content for verification.
+        Sha256 string `json:"sha256,omitempty"`
+
+        // ArtifactPullSecrets holds a reference to a secret where Docker or Basic Auth
+        // credentials are stored. We use these credentials when pulling the artifacts from
+        // the URL.
+        ArtifactPullSecret *ArtifactPullSecret `json:"artifactPullSecret,omitempty"`
```

Comment:
Am I assuming correct that his would then apply to the existing http protocol too? If so, maybe write it up a bit

Reply:
Yes, you are correct. This is the change I have introduced, let me know if it suffices: https://github.com/k0sproject/k0s/pull/4524/commits/89335cb350296a3bff5aeccaa7316f7c63b793fc